### PR TITLE
Add media_mime_type keyword argument for media upload methods

### PR DESCRIPTION
**Pull request by bjwatson**
Allows users to explicitly specify MIME type for media when we cannot auto-detect. We depend on the Python `mimetypes` module to do this, and it doesn't know about all MIME types (e.g. `audio/x-raw`).

### CODE REVIEW

File: googleapiclient/discovery.py [hunk 1]
```diff
@@ -774,9 +774,21 @@ def method(self, **kwargs):
     if media_filename:
       # Ensure we end up with a valid MediaUpload object.
       if isinstance(media_filename, six.string_types):
-        (media_mime_type, encoding) = mimetypes.guess_type(media_filename)
+        # FIXME: get this from kwargs['media_mime_type'], rather than mucking with body
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/discovery.py]: `TODO(<issue link>):` is preferable to `FIXME`.

File: googleapiclient/discovery.py [hunk 2]
```diff
@@ -774,9 +774,21 @@ def method(self, **kwargs):
...
+        media_mime_type = None
+        if body_value and body_value['metadata']:
+          metadata = body_value['metadata']
+          for key in metadata:
+            if key.lower() == 'content-type':
+              media_mime_type = metadata[key]
+              break
+
+        if media_mime_type is None:
+          (media_mime_type, encoding) = mimetypes.guess_type(media_filename)
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/discovery.py]: Drop these unnecessary parentheses.

File: googleapiclient/discovery.py (hunk 1 above)
Comment (jonparrott) [inline googleapiclient/discovery.py]: Why not go ahead and add a kwarg and fallback to `guess_type` if it's not specified (and issue a warning)?

Comment (bjwatson) [inline googleapiclient/discovery.py]: @jonparrott That's what I was thinking. I discussed this internally on the customer issue. Are you proposing a warning if the `media_mime_type` is not explicitly defined? That might be annoying for existing users given that this is a new feature, and they've always relied on auto-detect before.

@nathanielmanistaatgoogle I use `FIXME` deliberately when I want to address something _prior_ to merge, whereas `TODO` is for something to be fixed at some point in the future (post-merge). In a large PR, I can `git grep FIXME` to make sure I didn't miss anything.

File: googleapiclient/discovery.py (hunk 2 above)
Comment (bjwatson) [inline googleapiclient/discovery.py]: Ok. This line was pre-existing. All I did was indent it, but I can do that stylistic change.

**Comment by bjwatson** [comment]
@nathanielmanistaatgoogle I'm working on test coverage, but want to nail down how this is specified (i.e. using keyword args rather than body metadata).

File: googleapiclient/discovery.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline googleapiclient/discovery.py]: Makes sense, thanks.

Comment (bjwatson) [inline googleapiclient/discovery.py]: Ok, will do. :)

**Comment by bjwatson** [comment]
@nathanielmanistaatgoogle @jonparrott PTAL

File: googleapiclient/discovery.py [hunk 3]
```diff
@@ -774,7 +782,9 @@ def method(self, **kwargs):
     if media_filename:
       # Ensure we end up with a valid MediaUpload object.
       if isinstance(media_filename, six.string_types):
-        (media_mime_type, encoding) = mimetypes.guess_type(media_filename)
+        if media_mime_type is None:
+          logger.warning('media_mime_type argument undefined: trying to auto-detect for %s', media_filename)
```
Comment (jonparrott) [inline googleapiclient/discovery.py]: s/undefined/not specified

Comment (bjwatson) [inline googleapiclient/discovery.py]: Fixed

**Comment by bjwatson** [comment]
@nathanielmanistaatgoogle PTAL. This is blocked on you releasing your "Changes requested" flag.

**Comment by nathanielmanistaatgoogle** [comment]
Content looks good; squash commits and conform to [the guidelines](http://chris.beams.io/posts/git-commit/#seven-rules).

**Comment by bjwatson** [comment]
Normally we do that as part of the "Squash and merge button" in the googleapis org. Do you want me to do that manually on this branch first before you'll approve?

**Comment by bjwatson** [comment]
@nathanielmanistaatgoogle Is the "squash and merge" button acceptable, or do you want me to use `git rebase -i` so you can review my squashed commit message.

**Comment by nathanielmanistaatgoogle** [comment]
I like to review the squashed messages, especially now that "rebase and merge" is available as a final fire-the-missiles option.

**Comment by bjwatson** [comment]
@nathanielmanistaatgoogle Done.

**Pull request merged**